flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.ttl;
+
+import org.apache.flink.api.common.state.AggregatingStateDescriptor;
+import org.apache.flink.api.common.state.FoldingStateDescriptor;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.state.MapStateDescriptor;
+import org.apache.flink.api.common.state.ReducingStateDescriptor;
+import org.apache.flink.api.common.state.State;
+import org.apache.flink.api.common.state.StateDescriptor;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.flink.api.common.typeutils.CompositeSerializer;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.base.LongSerializer;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.state.KeyedStateFactory;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This state factory wraps state objects, produced by backends, with TTL logic.
+ */
+public class TtlStateFactory {
+	public static <N, SV, S extends State, IS extends S> IS createStateAndWrapWithTtlIfEnabled(
+		TypeSerializer<N> namespaceSerializer,
+		StateDescriptor<S, SV> stateDesc,
+		KeyedStateFactory originalStateFactory,
+		TtlConfig ttlConfig,
+		TtlTimeProvider timeProvider) throws Exception {
+		return ttlConfig.getTtlUpdateType() == TtlUpdateType.Disabled ?
+			originalStateFactory.createState(namespaceSerializer, stateDesc) :
+			new TtlStateFactory(originalStateFactory, ttlConfig, timeProvider)
+				.createState(namespaceSerializer, stateDesc);
+	}
+
+	private final Map<Class<? extends StateDescriptor>, StateFactory> stateFactories;
+
+	private final KeyedStateFactory originalStateFactory;
+	private final TtlConfig ttlConfig;
+	private final TtlTimeProvider timeProvider;
+
+	private TtlStateFactory(KeyedStateFactory originalStateFactory, TtlConfig ttlConfig, TtlTimeProvider timeProvider) {
```

Comment:
Would be better to check the args are not null, or simply use the `@Nonnull` annotation.

Reply:
I added checks in `createStateAndWrapWithTtlIfEnabled`, the only user of this private constructor.